optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/score/descriptor/ScoreDescriptor.java

```diff
@@ -0,0 +1,221 @@
+                throw new IllegalArgumentException("The solutionClass (" + solutionClass
+                        + ") has a " + PlanningScore.class.getSimpleName()
+                        + " annotated member (" + scoreMemberAccessor
+                        + ") that returns a scoreType (" + scoreType
+                        + ") that is not recognized as a default " + Score.class.getSimpleName() + " implementation.\n"
+                        + "  If you intend to use a custom implementation,"
+                        + " maybe set a scoreDefinition in the " + PlanningScore.class.getSimpleName()
+                        + " annotation.");
+            }
+        } else {
+            int bendableHardLevelsSize = annotation.bendableHardLevelsSize();
+            int bendableSoftLevelsSize = annotation.bendableSoftLevelsSize();
+            if (bendableHardLevelsSize == PlanningScore.NO_LEVEL_SIZE
+                    || bendableSoftLevelsSize == PlanningScore.NO_LEVEL_SIZE) {
+                throw new IllegalArgumentException("The solutionClass (" + solutionClass
+                        + ") has a " + PlanningScore.class.getSimpleName()
+                        + " annotated member (" + scoreMemberAccessor
+                        + ") that returns a scoreType (" + scoreType
+                        + ") that must have a bendableHardLevelsSize (" + annotation.bendableHardLevelsSize()
+                        + ") and a bendableSoftLevelsSize (" + annotation.bendableSoftLevelsSize() + ").");
+            }
+            if (scoreType.equals(BendableScore.class)) {
+                return new BendableScoreDefinition(bendableHardLevelsSize, bendableSoftLevelsSize);
+            } else if (scoreType.equals(BendableLongScore.class)) {
+                return new BendableLongScoreDefinition(bendableHardLevelsSize, bendableSoftLevelsSize);
+            } else if (scoreType.equals(BendableBigDecimalScore.class)) {
+                return new BendableBigDecimalScoreDefinition(bendableHardLevelsSize, bendableSoftLevelsSize);
+            } else {
+                throw new IllegalArgumentException("The solutionClass (" + solutionClass
+                        + ") has a " + PlanningScore.class.getSimpleName()
+                        + " annotated member (" + scoreMemberAccessor
+                        + ") that returns a bendable scoreType (" + scoreType
+                        + ") that is not recognized as a default " + Score.class.getSimpleName() + " implementation.\n"
+                        + "  If you intend to use a custom implementation,"
+                        + " maybe set a scoreDefinition in the annotation.");
+            }
+        }
+    }
+
+    public ScoreDefinition<?> getScoreDefinition() {
+        return scoreDefinition;
+    }
+
+    public Class<? extends Score<?>> getScoreClass() {
+        return scoreDefinition.getScoreClass();
+    }
+
+    public <Solution_> Score<?> getScore(Solution_ solution) {
```

Comment:
Consider perhaps generifying this class in <Solution_> too. There are more places than just this method here where it could be useful.

(Many classes throghout the solver are generic both in score and in solution.)

Reply:
`<Solution_>` here is an automatic refactoring error that I have overlooked.

My first choice would be simply `Object`. That doesn't hurt type safety neither inside nor outside of the `ScoreDescriptor`.

On second thought I could imagine we generify `ScoreDescriptor<Solution_>` just because of the convention that everything that deals with solution is generic. Would you prefer that knowing that `<Solution_>` parameter doesn't add any type safety?

Reply:
I gave it a try and I definitely downvote `ScoreDescriptor<Solution_>`. I found zero benefit.

The reason is that the majority of solution occurrences in this class is `solutionClass` that is used for constructing exception messages, so `Class<?> solutionClass` is perfectly OK, I think.

The rest of occurrences is in `getScore()/setScore()` where the solution type is irrelevant so I can't see the benefit of `Solution_` here either.

Reply:
Agree on `Solution_`.